### ISSUE TITLE
install.c/boot proposal

### ISSUE THREAD

**Issue by stsp** [stsp/dosemu2#302]
Somehow we need to start merging the
DOS installer, that is, to the date, barely
tested and not reviewed yet.
To simplify the task, here's the proposal:

1. $_bootdrive should default to (auto), which will
search for the bootable drive, starting from C. We
already have all API for that - trivial task.

2. Stop install.c from touching drive C at all.
After 1 is done, we can let the drive C to be empty
and dosemu will boot from D that contains freedos.
That reminds me the drive_z idea of dosemu1 -
history is always repeating, but this time without
any fatfs hacks and fully controllable with config.

3. Make D a synthetic symlink (as currently E is)
that will follow 'prefix' to locate the system-wide
DOS install. This will be possible because after 2
is implemented, there would be no symlinks to D.

This proposal will allow:
- merging the DOS installer easily, as install.c will
not touch C and installer will not touch anything
other than C.
- preparing the new supplementary packages, like
dosemu-freedos, dosemu-gnudos etc, that are in
'Suggests', i.e. no hard dep on them and no even
'Recommends'. Such packages can be in PPA forever
because they only add some extra functionality with
the DOS installer doing the main thing.

**Comment by stsp** [comment]
Hi @jschwartzenberg, sorry for this took so long.
Now drive C only contains the 2 symlinks: `command.com`
and `fdconfig.sys`.
Your install scripts should just unlink `command.com`
and replace it with another one.
So could you please update your scripts and open
the PR? Or what was the agreement? I don't remember.

**Comment by jschwartzenberg** [comment]
I don't remember either. I can do anything that would make sense, but do you think this will still be needed once FDPP is finished?

**Comment by stsp** [comment]
FDPP is just the kernel. It doesn't change the
fact that we still support every DOS. FDPP is
just a good starting point here, as now I have
a play-ground of 2 DOSes, so I am changing
the installation layout to match the new
requirements. This brings us much closer to
your task. So when changing the layout, it would
be nice to have your work handy so that to not
change the layout later once again for it.
FDPP is nothing without the freedos/gnudos tools,
so even with fdpp, freedos should still be installed.
FDPP doesn't yet even have its own command.com.

Did you think dosemu2 is going to support only
fdpp? Even dosbox supports other DOSes besides
its own, and I am sure dosbox does many things
rightly.
By the way, does dosbox offer the redirector
capabilities for "other DOSes", or only for its own?

**Comment by jschwartzenberg** [comment]
I think dosbox just boots other systems from images, but I'm not sure. Whether you would like something like this to install other DOSs when it would be optional to use them is a choice.

I'll have to dive in the code a bit again to see what it was doing. I guess I'd better start with rebasing them against the latest developments.

Note that a lot of software relies on DOS being in c:\\ like the Windows installer.

**Comment by stsp** [comment]
> I think dosbox just boots other systems from images, but I'm not sure.

But the booting procedure doesn't matter
because you may have image on an A and
a redirector on C. So the booted from image
DOS can still redirect.

> Whether you would like something like this to install other DOSs
> when it would be optional to use them is a choice.

Freedos can't be optional because of its tools.
fdpp does not reimplement tools.
So its only theoretically optional, but in reality
the user will have to be presented with the
install choice on a first start.


> Note that a lot of software relies on DOS being in c:\ like the Windows installer.

How do you mean? Do they expect some
tools in `C:\DOS` instead of relying on `PATH`,
or what?

**Comment by stsp** [comment]
So its not DOS but startup files.
The plan was to use `CHAIN` and all that
to process user's files on boot.
But this is a low-prio long-term plan.
Also it may be good to avoid user's
config.sys because the drivers may
be incompatible with dosemu. So
this should be thought very carefully
with adding config switches etc. But
its solvable.

**Comment by jschwartzenberg** [comment]
Alright, I'll ignore this for now then. Possible the default files will already be suitable for Windows. I'm not sure when I can look at the code again, but I'll try to find a moment for it!

**Comment by stsp** [comment]
> Alright, I'll ignore this for now then.

Right, as IIRC your scripts were aimed for
the host-side installing. So it doesn't matter
what the real install progs do.

**Comment by jschwartzenberg** [comment]
There is still the bug that nothing from my scripts was printed inside the dosemu window. Hopefully I can find the solution for that, but I can imagine you might figure that one out faster than me ;)

**Comment by stsp** [comment]
Yes, please present them as the PR, and
I'll try to take a look. I think the reason why
it doesn't print, was already explained by me,
but I don't remember. If there is an URL,
please provide.

**Comment by jschwartzenberg** [comment]
From what I remember is that we couldn't figure out the issue back then as I tried your suggestions but they did not change things.
Most if not all of what I did is currently in a branch in your fork already I think. Otherwise it's in mine. I'll provide a PR when there are new developments of course!

**Comment by ecm-pushbx** [comment]
> Did you think dosemu2 is going to support only
> fdpp? Even dosbox supports other DOSes besides
> its own, and I am sure dosbox does many things
> rightly.
> By the way, does dosbox offer the redirector
> capabilities for "other DOSes", or only for its own?

Only its own built-in DOS has the filesystem access in DOSBox, there is no redirector interfacing when booting an actual DOS. dosemu's redirector support is actually one of its greatest benefits.

**Comment by stsp** [comment]
Good, thanks for info.
Unless someone knows the otherwise, I would
assume the same applies to their IPX networking
and the rest.

> one of its greatest benefits.

Another one should be speed, of course.
